File: src/hackerrank/dataStructures/stacks/BalancedBrackets.java
package hackerrank.dataStructures.stacks;

import java.util.Stack;

public class BalancedBrackets {

    //https://www.hackerrank.com/challenges/balanced-brackets/

    public static void main(String[] args) {
        System.out.println(isBalanced("{{)[](}}"));
    }

    static String isBalanced(String s) {

        Stack<Character> stack = new Stack<>();

        for (Character c : s.toCharArray()) {
            addBracket(stack, c);
        }

        if (stack.isEmpty()) return "YES";
        else return "NO";
    }

    static void addBracket(Stack<Character> stack, Character c) {
        if (stack.isEmpty()) stack.add(c);
        else {
            Character stackTip = stack.pop();
            if (!areTheseAreOpposite(stackTip, c)) {
                stack.add(stackTip);
                stack.add(c);
            }
        }
    }

    static boolean areTheseAreOpposite(Character c1, Character c2) {
        return c1.equals('(') && c2.equals(')') ||
                c1.equals('{') && c2.equals('}') ||
                c1.equals('[') && c2.equals(']');
    }
}

File: src/kotlinCode/dataStructures/stacks/BalancedBrackets.kt
package kotlinCode.dataStructures.stacks

//https://www.hackerrank.com/challenges/balanced-brackets

fun isBalanced(s: String): String {
    val stringLength = s.length

    if (stringLength % 2 != 0) {
        return "NO"
    }

    var originalString = s

    while (
        originalString.contains("()") ||
        originalString.contains("[]") ||
        originalString.contains("{}")
    ) {
        originalString = originalString
            .replace("()", "")
            .replace("[]", "")
            .replace("{}", "")
    }

    if (originalString.isEmpty()) {
        return "YES"
    }

    return "NO"
}

fun main() {
    println(isBalanced("{(([])[])[]}")) //YES
    println(isBalanced("{(([])[])[]]}")) //NO
    println(isBalanced("{(([])[])[]}[]")) //YES

}